### PR TITLE
[IMP][12.0] stock_move_location: add posibility empty location with stock.picking

### PR DESCRIPTION
**Pull request by i-vyshnevska**
moved from https://github.com/OCA/stock-logistics-warehouse/pull/882
Addons add a button which allows quickly fill stock picking with everything in a bin

### CODE REVIEW

File: stock_move_location/models/stock_picking.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+                        'product_id': quant.product_id.id,
+                        'name': quant.product_id.partner_ref,
+                        'product_uom_qty': qty,
+                        'product_uom': quant.product_uom_id.id,
+                        'picking_type_id': self.picking_type_id.id,
+                        'location_id': self.location_id.id,
+                        'location_dest_id': self.location_dest_id.id,
+                    }
+                else:
+                    products[quant.product_id.id]['product_uom_qty'] += qty
```
Comment (mmequignon) [inline stock_move_location/models/stock_picking.py]: What's the purpose here, as qty should be 0 ?

**Review by mmequignon — APPROVED** [review]
Non-blocking comment, looks good

File: stock_move_location/models/stock_picking.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+from odoo import _, models
+from odoo.exceptions import UserError
+from itertools import groupby
+
+
+class StockPicking(models.Model):
+    _inherit = 'stock.picking'
+
+    def button_fillwithstock(self):
```
Comment (rousseldenis) [inline stock_move_location/models/stock_picking.py:12]: @i-vyshnevska Why not using the existing wizard to compute qunatities ? IMHO this has duplicate code from wizard one. Maybe you could do something more smart that works in both contexts no ?

File: stock_move_location/models/stock_picking.py (hunk 1 above)
Comment (i-vyshnevska) [reply stock_move_location/models/stock_picking.py]: if there is no qty that means we don't need to add this product to pikcing

File: stock_move_location/models/stock_picking.py (hunk 2 above)
Comment (i-vyshnevska) [reply stock_move_location/models/stock_picking.py:12]: @rousseldenis I have updated the code and reuse wizard features, but also I was needed to refactor wizard a little bit to fullfill our needs

Comment (i-vyshnevska) [reply stock_move_location/models/stock_picking.py:12]: and I have no idea why travis fails

File: stock_move_location/models/stock_picking.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        # check source location has no children, i.e. we scanned a bin
+
+        self.ensure_one()
+        if self.location_id.child_ids:
+            raise UserError(_('Please choose a source end location'))
+        if self.move_lines:
+            raise UserError(_('Moves lines already exists'))
+        quants = self.env['stock.quant'].search(
+            [
```
Comment (rousseldenis) [inline stock_move_location/models/stock_picking.py]: Just last thing. Could you put domain retrieval in a separate function allowing override ? Same for create (a _prepare function). Thanks

Comment (i-vyshnevska) [reply stock_move_location/models/stock_picking.py]: thanks, added with git amend

**Review by TDu — APPROVED** [review]

**Review by Tonow-c2c — APPROVED** [review]

File: stock_move_location/models/stock_picking.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        })
+        move_wizard._onchange_destination_location_id()
+        move_wizard.action_move_location()
+
+        return True
+
+    def _get_movable_quants(self):
+        return self.env['stock.quant'].search(
+            [
+                ('location_id', 'child_of', self.location_id.id),
```
Comment (sebalix) [inline stock_move_location/models/stock_picking.py]: Isn't this `child_of` misleading as we are checking just above that the location has no children? Perhaps use `=` instead?

File: stock_move_location/tests/test_stock_fillwithstock.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+            'location_id': self.shelf1_location.id,
+            'location_dest_id': self.pack_location.id,
+            'picking_type_id': self.env.ref('stock.picking_type_internal').id,
+        })
+        self.assertFalse(picking_stock_pack.move_lines)
+        picking_stock_pack.button_fillwithstock()
+        # picking filled with quants in bin
+        self.assertEqual(len(picking_stock_pack.move_lines), 2)
+        self.assertEqual(
+            picking_stock_pack.move_lines[0].product_id,
```
Comment (sebalix) [inline stock_move_location/tests/test_stock_fillwithstock.py]: We can't be sure that the first move will be the one related to `product1`, it is a bit wobbly I think. You should filter `move_lines` to get the relevant move and check its related product.

Comment (i-vyshnevska) [reply stock_move_location/tests/test_stock_fillwithstock.py]: @sebalix thanks, updated

**Comment by rousseldenis** [comment]
@i-vyshnevska Is this ready ?

**Comment by dreispt** [comment]
@rousseldenis Green build, no needs fixing - this looks good to merge.

**Review by marcelsavegnago — APPROVED** [review]

**Review by HaraldPanten — APPROVED** [review]

**Review by sebalix — APPROVED** [review]

**Comment by yvaucher** [comment]
Merge failed on 

```
odoo.addons.stock_pull_list.tests.test_stock_pull_list: ERROR: test_01_default_options (odoo.addons.stock_pull_list.tests.test_stock_pull_list.TestStockPullList)
odoo.addons.stock_pull_list.tests.test_stock_pull_list: Traceback (most recent call last):
odoo.addons.stock_pull_list.tests.test_stock_pull_list: `   File "/home/travis/build/OCA/stock-logistics-warehouse/stock_pull_list/tests/test_stock_pull_list.py", line 12, in test_01_default_options
odoo.addons.stock_pull_list.tests.test_stock_pull_list: `     wiz.action_prepare()
odoo.addons.stock_pull_list.tests.test_stock_pull_list: `   File "/home/travis/build/OCA/stock-logistics-warehouse/stock_pull_list/wizards/stock_pull_list_wizard.py", line 177, in action_prepare
odoo.addons.stock_pull_list.tests.test_stock_pull_list: `     key, demand_qty, supply_qty)))
odoo.addons.stock_pull_list.tests.test_stock_pull_list: `   File "/home/travis/build/OCA/stock-logistics-warehouse/stock_pull_list/wizards/stock_pull_list_wizard.py", line 94, in _prepare_line_values
odoo.addons.stock_pull_list.tests.test_stock_pull_list: `     rule = self._get_stock_rule_id(product, location)
odoo.addons.stock_pull_list.tests.test_stock_pull_list: `   File "/home/travis/build/OCA/stock-logistics-warehouse/stock_pull_list/wizards/stock_pull_list_wizard.py", line 128, in _get_stock_rule_id
odoo.addons.stock_pull_list.tests.test_stock_pull_list: `     product_id, location_id, values)
odoo.addons.stock_pull_list.tests.test_stock_pull_list: `   File "/home/travis/build/OCA/stock-logistics-warehouse/procurement_auto_create_group/models/procurement_group.py", line 13, in _get_rule
odoo.addons.stock_pull_list.tests.test_stock_pull_list: `     result = super()._get_rule(product_id, location_id, values)
odoo.addons.stock_pull_list.tests.test_stock_pull_list: `   File "/home/travis/odoo-12.0/addons/stock/models/stock_rule.py", line 352, in _get_rule
odoo.addons.stock_pull_list.tests.test_stock_pull_list: `     domain_company = ['|', ('company_id', '=', False), ('company_id', 'child_of', values['company_id'].ids)]
odoo.addons.stock_pull_list.tests.test_stock_pull_list: ` AttributeError: 'int' object has no attribute 'ids'
```

Which seems unrelated to me. I'll try to rerun the merge.

**Pull request merged**